crates/ruff_python_formatter/src/string/docstring.rs

```diff
@@ -1537,53 +1575,224 @@ fn needs_chaperone_space(normalized: &NormalizedString, trim_end: &str) -> bool
+            if iter.peek().copied() != Some('\t') {
+                return Indent::TabSpaces {
+                    tabs,
+                    spaces: align_spaces,
+                };
+            }
+        } else if align_spaces == 0 {
+            return Indent::SpacesTabs { spaces, tabs };
+        }
+
+        // Sequence of spaces.. tabs, spaces, tabs...
+        let mut indentation = spaces + tabs * TAB_INDENT_WIDTH + align_spaces;
+
+        for char in iter {
+            if char == '\t' {
+                // Pad to the next multiple of tab_width
+                indentation += TAB_INDENT_WIDTH - (indentation.rem_euclid(TAB_INDENT_WIDTH));
+            } else if char.is_whitespace() {
+                indentation += char.len_utf8();
+            } else {
+                break;
+            }
         }
+
+        // Mixed tabs and spaces
+        Indent::Mixed(indentation)
     }
-    indentation
-}
 
-/// Trims at most `indent_len` indentation from the beginning of `line`.
-///
-/// This treats indentation in precisely the same way as `indentation_length`.
-/// As such, it is expected that `indent_len` is computed from
-/// `indentation_length`. This is useful when one needs to trim some minimum
-/// level of indentation from a code snippet collected from a docstring before
-/// attempting to reformat it.
-fn indentation_trim(indent_len: usize, line: &str) -> &str {
-    let mut seen_indent_len = 0;
-    let mut trimmed = line;
-    for char in line.chars() {
-        if seen_indent_len >= indent_len {
-            return trimmed;
+    /// Returns the indentation's visual width in columns/spaces.
+    ///
+    /// For docstring indentation, black counts spaces as 1 and tabs by increasing the indentation up
+    /// to the next multiple of 8. This is effectively a port of
+    /// [`str.expandtabs`](https://docs.python.org/3/library/stdtypes.html#str.expandtabs),
+    /// which black [calls with the default tab width of 8](https://github.com/psf/black/blob/c36e468794f9256d5e922c399240d49782ba04f1/src/black/strings.py#L61).
+    const fn len(self) -> usize {
+        match self {
+            Indent::Spaces(count) => count,
+            Indent::Tabs(count) => count * TAB_INDENT_WIDTH,
+            Indent::TabSpaces { tabs, spaces } => tabs * TAB_INDENT_WIDTH + spaces,
+            Indent::SpacesTabs { spaces, tabs } => {
+                let mut indent = spaces;
+                indent += TAB_INDENT_WIDTH - indent.rem_euclid(TAB_INDENT_WIDTH);
+                indent + (tabs - 1) * TAB_INDENT_WIDTH
+            }
+            Indent::Mixed(width) => width,
         }
-        if char == '\t' {
-            // Pad to the next multiple of tab_width
-            seen_indent_len += 8 - (seen_indent_len.rem_euclid(8));
-            trimmed = &trimmed[1..];
-        } else if char.is_whitespace() {
-            seen_indent_len += char.len_utf8();
-            trimmed = &trimmed[char.len_utf8()..];
+    }
```

Comment:
If this is returning the visual width, maybe a name other than `len` would be appropriate here? I usually expect a value returned by a method name `len()` to be usable as part of a range/offset calculation. `visual_len`? `visual_width`? `number_of_spaces`? `columns`?

Reply:
I like it. I think I'll go with `width`. It aligns nicely with `indent_width.` It then reads as `indentation.width()`.

Reply:
Oh yeah that's great!